I'm flying blind on these units tests, as I still can't compile and run due to dependency issues, but hopefully they should work...

```
>npm i
npm WARN ERESOLVE overriding peer dependency
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: @remix-run/dev@2.2.0
npm ERR! Found: typescript@4.9.5
npm ERR! node_modules/typescript
npm ERR!   peerOptional typescript@"^5.1.0" from @remix-run/node@2.2.0
npm ERR!   node_modules/@remix-run/dev/node_modules/@remix-run/node
npm ERR!     @remix-run/node@"2.2.0" from @remix-run/dev@2.2.0
npm ERR!     node_modules/@remix-run/dev
npm ERR!       dev @remix-run/dev@"^2.0.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peerOptional typescript@"^5.1.0" from @remix-run/dev@2.2.0
npm ERR! node_modules/@remix-run/dev
npm ERR!   dev @remix-run/dev@"^2.0.0" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: typescript@5.2.2
npm ERR! node_modules/typescript
npm ERR!   peerOptional typescript@"^5.1.0" from @remix-run/dev@2.2.0
npm ERR!   node_modules/@remix-run/dev
npm ERR!     dev @remix-run/dev@"^2.0.0" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

--- a/src/utils/__tests__/rate-limiter.test.ts
+++ b/src/utils/__tests__/rate-limiter.test.ts
@@ -1,3 +1,5 @@
+import { RateLimiter } from "../rate-limiter";
+
 describe('RateLimiter', () => {
   it('should not allow more than the rate limit to run concurrently', async () => {
     const limit = new RateLimiter(3);

--- a/src/utils/__tests__/rate-limiter.test.ts
+++ b/src/utils/__tests__/rate-limiter.test.ts
@@ -1,22 +1,29 @@
-import { RateLimiter } from "../rate-limiter";
+import { RateLimiter } from '../rate-limiter';
 
 describe('RateLimiter', () => {
   it('should not allow more than the rate limit to run concurrently', async () => {
     const limit = new RateLimiter(3);
     let activeTasks = 0;
-    const increaseActive = () => { activeTasks++; };
-    const decreaseActive = () => { activeTasks--; limit.free(); };
-    const tasks = Array(5).fill(null).map(async () => {
-      await limit.allocate();
+    const increaseActive = () => {
+      activeTasks++;
+    };
+    const decreaseActive = () => {
+      activeTasks--; limit.free();
+    };
+    const tasks = Array(5)
+      .fill(null).map(async () => {
+        await limit.allocate();
 
-      increaseActive();
-      expect(activeTasks).toBeLessThanOrEqual(3);
+        increaseActive();
+        expect(activeTasks).toBeLessThanOrEqual(3);
 
-       // Mock task duration (randomness to affect completion order)
-      await new Promise(resolve => setTimeout(resolve, 500 + 500*Math.random()));
-      decreaseActive();
-    });
-    await Promise.all(tasks);
+        // Mock task duration (randomness to affect completion order)
+        await new Promise(resolve =>
+          setTimeout(resolve, 500 + 500 * Math.random())
+        );
+        decreaseActive();
+      });
+      await Promise.all(tasks);
   });
 
   it('should queue tasks and execute them in order', async () => {
@@ -27,7 +34,7 @@ describe('RateLimiter', () => {
       taskOrder.push(id);
 
       // Mock task duration (randomness to affect completion order)
-      await new Promise(resolve => setTimeout(resolve, 500 + 500*Math.random()));
+      await new Promise(resolve => setTimeout(resolve, 500));
       limit.free();
     };
 
@@ -45,7 +52,7 @@ describe('RateLimiter', () => {
     };
 
     await Promise.all([mockTask(), mockTask()]);
-    // Check internal state or behavior to ensure it's handling this correctly
-    // This might depend on the implementation details of RateLimiter
+    expect(limit.getProcessing()).toEqual(0);
+    expect(limit.getWaiting()).toEqual(0);
   });
 });

--- a/src/utils/__tests__/rate-limiter.test.ts
+++ b/src/utils/__tests__/rate-limiter.test.ts
@@ -31,10 +31,11 @@ describe('RateLimiter', () => {
     let taskOrder = [];
     const mockTask = async (id) => {
       await limit.allocate();
-      taskOrder.push(id);
 
       // Mock task duration (randomness to affect completion order)
-      await new Promise(resolve => setTimeout(resolve, 500));
+      await new Promise(resolve => setTimeout(resolve, 500 + 500*id));
+
+      taskOrder.push(id);
       limit.free();
     };
 

--- a/src/utils/rate-limiter.ts
+++ b/src/utils/rate-limiter.ts
@@ -40,4 +40,11 @@ export class RateLimiter {
       this.active = Math.max(0, this.active - 1);
     }
   }
+
+  getProcessing(): number {
+    return this.active;
+  }
+  getWaiting(): number {
+    return this.queue.length;
+  }
 }

